hpms/src/main/java/gov/cms/ab2d/hpms/service/HPMSAuthServiceImpl.java

```diff
@@ -0,0 +1,93 @@
+package gov.cms.ab2d.hpms.service;
+
+import gov.cms.ab2d.hpms.hmsapi.HPMSAuthResponse;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Service;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Flux;
+
+import java.time.Duration;
+
+import static org.springframework.http.HttpHeaders.*;
+
+@Service
+public class HPMSAuthServiceImpl implements HPMSAuthService {
+
+    @Value("${hpms.auth.url}")
+    private String authURL;
+
+    @Value("${HPMS_AUTH_KEY_ID}")
+    private String hpmsAPIKeyId;
+
+    @Value("${HPMS_AUTH_KEY_SECRET}")
+    private String hpmsSecret;
+
+    private volatile String authToken;
+
+    private volatile long tokenExpires;
```

Comment:
Why do you need volatile here?

Reply:
This service is effectively a singleton with potentially multiple callers coming in on multiple threads.   Using volatile at least guarantees a memory latch which insures that each thread has an opportunity to see the latest state which may be a newly minted token.  While not guaranteeing multiple token requests, it makes it far more likely that there won't be.

Otherwise, different threads may not see the common state which can cause churning.

Reply:
I suppose that would be a possibility. Might be a bit of an edge case since I think an admin request would be the only caller

Reply:
At a service level, I tend not to assume how its called.  Volatile is very cheap (execution-wise) and calls attention to the potential threading issues.